lib/semantics/tools.cc

```diff
@@ -508,6 +509,169 @@ std::optional<parser::MessageFixedText> WhyNotModifiable(
   }
 }
 
+bool IsCoarray(const parser::AllocateObject &allocateObject) {
+  const parser::Name &name{GetLastName(allocateObject)};
+  return name.symbol && IsCoarray(*name.symbol);
+}
+
+bool IsImageControlStmt(const parser::AllocateStmt &allocateStmt) {
+  const auto &allocationList{
+      std::get<std::list<parser::Allocation>>(allocateStmt.t)};
+  for (const auto &allocation : allocationList) {
+    const auto &allocateObject{std::get<parser::AllocateObject>(allocation.t)};
+    if (IsCoarray(allocateObject)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool IsImageControlStmt(const parser::DeallocateStmt &deallocateStmt) {
+  const auto &allocateObjectList{
+      std::get<std::list<parser::AllocateObject>>(deallocateStmt.t)};
+  for (const auto &allocateObject : allocateObjectList) {
+    if (IsCoarray(allocateObject)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool IsImageControlStmt(const parser::CallStmt &callStmt) {
+  const auto &procedureDesignator{
+      std::get<parser::ProcedureDesignator>(callStmt.v.t)};
+  if (auto *name{std::get_if<parser::Name>(&procedureDesignator.u)}) {
+    // TODO: also ensure that the procedure is, in fact, an intrinsic
+    if (name->source == "move_alloc") {
+      const auto &args{
+          std::get<std::list<parser::ActualArgSpec>>(callStmt.v.t)};
+      if (!args.empty()) {
+        const parser::ActualArg &actualArg{
+            std::get<parser::ActualArg>(args.front().t)};
+        if (const auto *argExpr{
+                std::get_if<common::Indirection<parser::Expr>>(&actualArg.u)}) {
+          return HasCoarray(argExpr->value());
+        }
+      }
+    }
+  }
+  return false;
+}
+
+bool IsImageControlStmt(
+    const parser::Statement<parser::ActionStmt> &actionStmt) {
+  return std::visit(
+      common::visitors{
+          [](const common::Indirection<parser::EventPostStmt> &) {
```

Comment:
All of these `return true;` cases can be rolled into the `const auto &` default below and implemented with a `HasMember` test.

Reply:
Extending this idea, I think it would simplify the code to add an explicit visitor class to pass to the two calls to `std::visit`. This would encapsulate all of the code to implement `IsImageControlStatement` in one place.

Something like this:
```
struct ImageControlStmtHelper {
  using ImageControlStmts = std::variant<  ... >;
  template<typename T> bool operator()(const T &) {
    return common::HasMember<T, ImageControlStmts>;
  }
  template<typename T> bool operator()(const common::Indirection<T> &x) {
    return (*this)(x.value());
  }
  bool operator()(const parser::AllocateStmt &stmt) ...
  bool operator()(const parser::DeallocateStmt &stmt) ...
  bool operator()(const parser::CallStmt &stmt) ...
  bool operator()(const parser::Statement<parser::ActionStmt> &stmt) {
    return std::visit(*this, stmt.statement.u);
  }
};
bool IsImageControlStmt(const parser::ExecutableConstruct &construct) {
  return std::visit(ImageControlStmtHelper{}, construct.u);
}
```

Reply:
Thanks, Tim and Peter.  I'll make this change.